sfe/zendesk/zendesk.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	var query strings.Builder
+	query.WriteString("type:ticket")
+
+	for name, want := range matchFields {
+		id, ok := c.nameToFieldID[name]
+		if !ok {
+			return nil, fmt.Errorf("unknown custom field %q", name)
+		}
+		val := want
+		if strings.ContainsRune(val, ' ') {
+			val = `"` + val + `"`
+		}
+		fmt.Fprintf(&query, " custom_field_%d:%s", id, val)
+	}
+
+	searchURL := c.searchURL + "?query=" + url.QueryEscape(query.String())
+	out := make(map[int64]map[string]string)
+
+	for searchURL != "" {
+		req, err := c.newJSONRequest(http.MethodGet, searchURL, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create zendesk request: %w", err)
+		}
+
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode >= 300 {
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("zendesk status %d: failed to read response body: %w", resp.StatusCode, err)
+			}
+			resp.Body.Close()
+			return nil, fmt.Errorf("zendesk status %d: %s", resp.StatusCode, body)
+		}
+
+		var results struct {
+			Results []struct {
+				ID           int64 `json:"id"`
+				CustomFields []struct {
+					ID    int64 `json:"id"`
+					Value any   `json:"value"`
+				} `json:"custom_fields"`
+			} `json:"results"`
+			Next *string `json:"next_page"`
+		}
+		err = json.NewDecoder(resp.Body).Decode(&results)
+		if err != nil {
+			resp.Body.Close()
+			return nil, fmt.Errorf("failed to decode zendesk response: %w", err)
+		}
+		resp.Body.Close()
+
+		for _, result := range results.Results {
+			fieldMap := make(map[string]string)
+			for _, cf := range result.CustomFields {
+				name, ok := c.fieldIDToName[cf.ID]
+				if ok {
+					fieldMap[name] = fmt.Sprint(cf.Value)
+				}
+			}
+			out[result.ID] = fieldMap
+		}
+		if results.Next != nil {
+			searchURL = *results.Next
+			continue
+		}
+		searchURL = ""
+	}
+	return out, nil
+}
+
+// AddComment posts the comment body to the specified ticket. The comment is
+// added as a public or private comment based on the provided boolean value. An
+// error is returned if the request fails.
+func (c *Client) AddComment(ticketID int64, commentBody string, public bool) error {
+	endpoint, err := url.JoinPath(c.commentURL, fmt.Sprintf("%d.json", ticketID))
+	if err != nil {
+		return fmt.Errorf("failed to join ticket path: %w", err)
+	}
+
+	payload := struct {
+		Ticket struct {
+			Comment comment `json:"comment"`
+		} `json:"ticket"`
+	}{}
+	payload.Ticket.Comment = comment{Body: commentBody, Public: public}
+
+	body, err := json.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal zendesk comment: %w", err)
+	}
+
+	req, err := c.newJSONRequest(http.MethodPut, endpoint, body)
+	if err != nil {
+		return fmt.Errorf("failed to create zendesk request: %w", err)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("zendesk request failed: %w", err)
+	}
+	defer resp.Body.Close()
```

Comment:
Every call to newJSONRequest has these same lines after it. Should they be bundled into that function, to have it return a result object?

My first thought was that having the helper function just return the request object was a smart idea for unit-testing the helper, but I don't see any standalone unit tests for the helper.